compiler/optimizer/OMRSimplifierHandlers.cpp

```diff
@@ -5260,6 +5260,118 @@ static bool skipRemLowering(int64_t divisor, TR::Simplifier *s)
    }
 
 
+/**
+ * @brief Reduce Ifcmp Greater-Than-Or-Equal to Ifcmp Equal in a specific case.
+ *
+ * If
+* - the first child is an "and" node
+* - the second child of the "and" node is a constant N
+* - the second child of the ifcmpge node is a constant N,
+* Then, we can reduce ifcmpge to ifcmpeq.
+*
+* For example, convert
+* ificmpge --> block_5
+*  iand
+*    iload  <...>
+*    iconst 16
+*  iconst 16
+*
+*  into
+*
+* ificmpeq --> block_5
+*  iand
+*    iload  <...>
+*    iconst 16
+*  iconst 16
+*
+* Since the constants are equal, the case of greater-than is impossible,
+* and we can reduce ifcmpge to ifcmpeq. In other words, the result of iand
+* cannot be greater than the constant, and we can simply perform an equality check,
+* which provides further optimization opportunities.
+*/
+class IfcmpgeReducer
+   {
+   public:
+   IfcmpgeReducer(TR::Node *node, TR::Simplifier *s) : _node(node), _s(s) {}
+
+   /**
+    * @brief Get a specific ifcmpeq opcode based on a ifcmpge opcode.
+    *
+    * Each ifcmpge opcode is mapped to its corresponding ifcmpeq opcode.
+    * E.g., ifiucmpge should be substituted with ifiucmpeq.
```

Comment:
Equality is insensitive to signedness, so there's no issue changing `ifiucmpge` into `ificmpeq` (without the `u`).  Then this function would just be `TR::ILOpCodes::ifcmpeqOpCode(node->getDataType())`.  The `ucmpeq` and `ucmpne` opcodes should probably be considered deprecated, as with `uadd` and `usub`

Reply:
Thanks for your suggestion, Davin. I changed the function but now I'm getting this assertion when I run my tests:
```
Assertion failed at /sandbox/azhikhar/omr/compiler/il/OMRILOps.hpp:1053: 0

        no ifcmpeq opcode for this datatype NoType
```

I might be missing something. @Leonardo2718 is it possible that Tril does not set the data type for the node?

Reply:
Oh sorry, I should have said to take the data type of a child

Reply:
Ah, gotcha. Fixed.